sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -769,12 +675,68 @@ int gcoap_req_init(coap_pkt_t *pdu, uint8_t *buf, size_t len,
 
 ssize_t gcoap_finish(coap_pkt_t *pdu, size_t payload_len, unsigned format)
 {
-    /* reconstruct full PDU buffer length */
-    size_t len = pdu->payload_len + (pdu->payload - (uint8_t *)pdu->hdr);
+    unsigned format_offset = UINT_MAX;
+    unsigned len_delta = 0;     /* buffer length to move forward */
+    uint8_t *move_pos = 0;      /* buffer position to move forward */
+
+    /* Remove or adjust Content-Format option for request, if any.
+     * Assumes current payload format is COAP_FORMAT_NO_PAYLOAD */
+    for (unsigned i = 0; i < pdu->options_len; i++) {
+        if (pdu->options[i].opt_num == COAP_OPT_CONTENT_FORMAT) {
+            /* assume option value is COAP_FORMAT_NO_PAYLOAD */
+            format_offset = i;
+            if (pdu->options_len > (i + 1)) {
+                len_delta = pdu->options[i+1].offset - pdu->options[i].offset;
+            }
+            else {
+                len_delta = (pdu->payload - (uint8_t *)pdu->hdr)
+                                - pdu->options[i].offset;
+            }
+            if (format == COAP_FORMAT_NONE) {
+                pdu->options_len--;
+            }
+            else {
+                /* new format must be shorter; COAP_FORMAT_NO_PAYLOAD is 3 bytes */
+                len_delta -= coap_put_option_ct(
+                                (uint8_t *)pdu->hdr + pdu->options[i].offset,
+                                (i > 0) ? pdu->options[i-1].opt_num : 0, format);
+            }
+        }
+        else if (format_offset < UINT_MAX) {
+            if (format == COAP_FORMAT_NONE) {
+                if (!move_pos) {
+                    move_pos = (uint8_t *)pdu->hdr + pdu->options[i].offset;
+                }
```

Comment:
These snippet  of 3 lines is in both branches of the `if`. Is it possible to move it outside of the `if` statement

Reply:
I agree -- looks like an oversight.